### PR TITLE
R298 plfm 6108

### PR DESCRIPTION
**Pull request by xschildw**

### CODE REVIEW

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java [hunk 1]
```diff
@@ -3,73 +3,193 @@
...
-    			jiraServerUri, 
-    			StackConfigurationSingleton.singleton().getJiraUserEmail(), 
-    			StackConfigurationSingleton.singleton().getJiraUserApikey());
+		SimpleHttpClientConfig httpClientConfig = new SimpleHttpClientConfig();
+		httpClientConfig.setSocketTimeoutMs(TIME_OUT);
+		httpClient = new SimpleHttpClientImpl(httpClientConfig);
+
+		USERNAME = StackConfigurationSingleton.singleton().getJiraUserEmail();
+		APIKEY = StackConfigurationSingleton.singleton().getJiraUserApikey();
+		parser = new JSONParser();
```
Comment (xschildw) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java]: Use when neeeded, don't cache. Also check if parser is thread-safe.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java [hunk 2]
```diff
@@ -3,73 +3,193 @@
...
+
+		SimpleHttpRequest request = new SimpleHttpRequest();
+		if (resource == null) {
+			request.setUri(JIRA_URL + path);
+		} else {
+			request.setUri(JIRA_URL + path + resource);
+		}
+		Map<String, String> headers = new HashMap<>();
+		try {
+			headers.put(HttpHeaders.AUTHORIZATION, "Basic " + Base64.getEncoder().encodeToString((USERNAME + ":" + APIKEY).getBytes("utf-8")));
```
Comment (xschildw) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java]: make private method to et all headers

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java [hunk 3]
```diff
@@ -3,73 +3,193 @@
...
+		} catch (UnsupportedEncodingException e) {
+			throw new RuntimeException("Error encoding credentials");
+		}
+		if (contentType != null) {
```
Comment (xschildw) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java]: Always set content type to app/json. Check if constant exists somewhere.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java [hunk 4]
```diff
@@ -3,73 +3,193 @@
...
+			headers.put(HttpHeaders.CONTENT_TYPE, contentType);
+		}
+		headers.put(HttpHeaders.USER_AGENT, USER_AGENT);
+		request.setHeaders(headers);
 
+		return request;
+	}
+
+	static void handleResponseStatus(int status) {
```
Comment (xschildw) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java:170]: Create a dedicated Exception resulting in Bad Gateway.
Also surface the msg.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java [hunk 5]
```diff
@@ -3,73 +3,193 @@
...
+			throw new RuntimeException();
+		}
+		handleResponseStatus(resp.getStatusCode());
+		String json = null;
+		json = resp.getContent();
+
+		Long issueTypeId = -1L;
+		String projectId = null;
+		try (Reader reader = new StringReader(json)) {
+			JSONObject pInfo = (JSONObject) parser.parse(reader);
```
Comment (xschildw) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java]: parse() takes string so no need for reader

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java [hunk 6]
```diff
@@ -3,73 +3,193 @@
...
+					break;
+				}
+			}
+			if (issueTypeId == -1L) {
+				throw new IllegalStateException(String.format("Could not find the issue typeId for issue type {}", issueTypeName));
+			}
+			projectId = (String) pInfo.get(JIRA_PROJECT_ID_KEY);
+		}
+		catch (IOException | ParseException e) {
+			throw new RuntimeException("JIRA client: error processing JSON");
```
Comment (xschildw) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java]: IllegalState

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java [hunk 7]
```diff
@@ -3,73 +3,193 @@
...
+				String name = (String) issueType.get("name");
+				if (issueTypeName.equalsIgnoreCase(name)) {
+					issueTypeId = Long.parseLong((String) issueType.get("id"));
+					break;
+				}
+			}
+			if (issueTypeId == -1L) {
+				throw new IllegalStateException(String.format("Could not find the issue typeId for issue type {}", issueTypeName));
+			}
+			projectId = (String) pInfo.get(JIRA_PROJECT_ID_KEY);
```
Comment (xschildw) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java:76]: Handle null case

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java [hunk 8]
```diff
@@ -3,73 +3,193 @@
...
+		try {
+			resp = httpClient.get(req);
+		} catch (IOException e) {
+			throw new RuntimeException();
+		}
+		handleResponseStatus(resp.getStatusCode());
+		String json = null;
+		json = resp.getContent();
+
+		Long issueTypeId = -1L;
```
Comment (xschildw) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java]: Leave to null

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java [hunk 9]
```diff
@@ -3,73 +3,193 @@
...
+		String projectId = null;
+		try (Reader reader = new StringReader(json)) {
+			JSONObject pInfo = (JSONObject) parser.parse(reader);
+			JSONArray issueTypeNames = (JSONArray)pInfo.get(JIRA_PROJECT_ISSUE_TYPES_KEY);
+			Iterator<JSONObject> it = issueTypeNames.iterator();
+			while (it.hasNext()) {
+				JSONObject issueType = it.next();
+				String name = (String) issueType.get("name");
+				if (issueTypeName.equalsIgnoreCase(name)) {
+					issueTypeId = Long.parseLong((String) issueType.get("id"));
```
Comment (xschildw) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java]: use valueOf(), check that get() does not return null. And extract "id" as constant.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java [hunk 10]
```diff
@@ -3,73 +3,193 @@
...
+			}
+			if (issueTypeId == -1L) {
+				throw new IllegalStateException(String.format("Could not find the issue typeId for issue type {}", issueTypeName));
+			}
+			projectId = (String) pInfo.get(JIRA_PROJECT_ID_KEY);
+		}
+		catch (IOException | ParseException e) {
+			throw new RuntimeException("JIRA client: error processing JSON");
+		}
+		JSONObject projInfo = new JSONObject();
```
Comment (xschildw) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java:81]: Create a small DTO object.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java [hunk 11]
```diff
@@ -3,73 +3,193 @@
...
 	@Override
-	public Iterable<Field> getFields() {
-		// NOTE:  In order to appear in this list a Jira 'custom field' 
-		// must have a 'global context'.  If it applies just to selected
-		// projects it will not appear in this list.
-    	MetadataRestClient mrc = restClient.getMetadataClient();
-    	return waitForPromise(mrc.getFields());
+	public Map<String,String> getFields() {
+		SimpleHttpRequest req = createRequest(JIRA_API_FIELDS_URL, null, "application/json");
+		SimpleHttpResponse resp = null;
```
Comment (xschildw) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java]: Extract method

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java [hunk 12]
```diff
@@ -3,73 +3,193 @@
...
+		}
+		return fieldsMap;
 	}
 
 	@Override
-	public BasicIssue createIssue(IssueInput issueInput) {
-        IssueRestClient irc = restClient.getIssueClient();
-        Promise<BasicIssue> basicIssuePromise = irc.createIssue(issueInput);
-        return waitForPromise(basicIssuePromise);
+	public JSONObject createIssue(JSONObject issue) {
```
Comment (xschildw) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JiraClientImpl.java:113]: Create a DTO for issue

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/util/jrjc/JiraClientImplTest.java [hunk 13]
```diff
@@ -0,0 +1,340 @@
+package org.sagebionetworks.repo.util.jrjc;
+
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpStatus;
+import org.apache.http.ParseException;
+import org.json.simple.JSONObject;
```
Comment (xschildw) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/util/jrjc/JiraClientImplTest.java:6]: Move to JUnit5

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JRJCHelper.java [hunk 14]
```diff
@@ -47,41 +48,28 @@ public static String createRestrictIssue(JiraClient jiraClient, String principal
...
-			if (issueTypeName.equalsIgnoreCase(it.getName())) issueTypeId = it.getId();
-		}
-		if (issueTypeId==-1L) throw new IllegalStateException("Cannot find issue type "+issueTypeName+" in Jira project "+JIRA_PROJECT_KEY);
-		
-		// third, find the defined fields, mapping their names to their IDs	
-		Map<String, String> lcFieldNameToIdMap = new HashMap<String, String>();
-		for (Field field : jiraClient.getFields()) {
-			lcFieldNameToIdMap.put(field.getName().toLowerCase(), field.getId());
-		}
+	public static JSONObject createIssue(JiraClient jiraClient, String issueTypeName, String summary, Map<String,String> params) {
```
Comment (xschildw) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/util/jrjc/JRJCHelper.java:44]: Create a DTO and push most of this inside.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/util/jrjc/JRJCHelperTest.java [hunk 15]
```diff
@@ -1,92 +1,98 @@
 package org.sagebionetworks.repo.util.jrjc;
```
Comment (xschildw) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/util/jrjc/JRJCHelperTest.java:1]: Convert to junit5

**Review by marcomarasca — APPROVED** [review]
Ok to merge for now so we can deploy and test in staging. Will need to be revised.

**Pull request merged**